client/image.go

```diff
@@ -306,6 +319,23 @@ func (i *image) Unpack(ctx context.Context, snapshotterName string, opts ...Unpa
 		return err
 	}
 
+	if !config.DisableSnapshotAnnotations {
+		var manifestDigest string
+		if manifest.Annotations != nil {
+			manifestDigest = manifest.Annotations[images.ManifestDigestLabel]
```

Comment:
Should be added from the client.

Reply:
The image handler needs descriptor of manifest, but `i.getManifest(ctx, i.platform)` doesn't return it.  I create the descriptor and set the digest passed by an annotation added in method `getManifest` .

Reply:
Do I have any other ways to get this descriptor?

Reply:
I meant this entire `if` block. According to your comment `the additional annotations are not passed to the snapshotter, when unpacking the local image`, the problem is that the annotation handler can't be used for importing local images, right? If so, how about adding a `client.ImportOpt` for receiving a handler wrapper (just like `client.WithImageHandlerWrapper`)?

Reply:
`client.WithImageHandlerWrapper`  will only be executed when pulling the remote image, and will not be executed when unpacking local image. 
When calling the `CreateContainer ` of `cri`,  if the specified image exists locally but no snapshot, it will call `Unpack ` method of `Image`, where no handler wrapper  will be configured and executed.
```
// WithNewSnapshot wraps `containerd.WithNewSnapshot` so that if creating the
// snapshot fails we make sure the image is actually unpacked and retry.
func WithNewSnapshot(id string, i containerd.Image, opts ...snapshots.Opt) containerd.NewContainerOpts {
	f := containerd.WithNewSnapshot(id, i, opts...)
	return func(ctx context.Context, client *containerd.Client, c *containers.Container) error {
		if err := f(ctx, client, c); err != nil {
			if !errdefs.IsNotFound(err) {
				return err
			}

			if err := i.Unpack(ctx, c.Snapshotter); err != nil {
				return fmt.Errorf("error unpacking image: %w", err)
			}
			return f(ctx, client, c)
		}
		return nil
	}
}
```

Reply:
In my scenario, two runtime handlers are configured. When I create first pod using the first runtime handler in Kubernetes, the additional annotations can be successfully passed to the snapshotter, because `PullImage ` of `cri ` is called, and an `AppendInfoHandlerWrapper ` will be added here.  However, when creating the second pod (whose image is the same as the previous one) with another runtime handler, the additional annotations are not passed to the snapshotter. Because the image already exists locally, the `PullImage ` method is not called but the `Unpack ` of the `Image ` is called instead, but the code of the handler is not executed in the unpack method.
The following is the key code in PullImage：
```golang
// Temporarily removed for v2 upgrade
	//pullOpts = append(pullOpts, c.encryptedImagesPullOpts()...)
	if !c.config.DisableSnapshotAnnotations {
		pullOpts = append(pullOpts,
			containerd.WithImageHandlerWrapper(snpkg.AppendInfoHandlerWrapper(ref)))
	}

	if c.config.DiscardUnpackedLayers {
		// Allows GC to clean layers up from the content store after unpacking
		pullOpts = append(pullOpts,
			containerd.WithChildLabelMap(containerdimages.ChildGCLabelsFilterLayers))
	}

	pullReporter.start(pctx)
	image, err := c.client.Pull(pctx, ref, pullOpts...)
```